File: src/models/frontend/Condition/Condition.ts
import { Exclude, instanceToPlain, plainToInstance } from 'class-transformer';
import { db_Condition } from 'models/db/db_Condition';
import { AffectedTraits } from 'models/frontend/Phenotype/Phenotype';

interface ICondition extends AffectedTraits {
  name: string;
  description?: string;
}

export class Condition {
  name: string = '';
  description?: string;
  maleMating?: number;
  lethal?: boolean;
  femaleSterile?: boolean;
  arrested?: boolean;
  maturationDays?: number;

  constructor(fields: ICondition) {
    Object.assign(this, fields);
  }

  static createFromRecord(record: db_Condition): Condition {
    return new Condition({
      name: record.name,
      description: record.description ?? undefined,
      maleMating:
        record.maleMating !== null ? Number(record.maleMating) : undefined,
      lethal: record.lethal ?? undefined,
      femaleSterile: record.femaleSterile ?? undefined,
      arrested: record.arrested ?? undefined,
      maturationDays: record.maturationDays ?? undefined,
    });
  }

  @Exclude()
  public generateRecord(): db_Condition {
    return {
      name: this.name,
      description: this.description ?? null,
      maleMating: this.maleMating !== undefined ? this.maleMating : null,
      lethal: this.lethal ?? null,
      femaleSterile: this.femaleSterile ?? null,
      arrested: this.arrested ?? null,
      maturationDays: this.maturationDays ?? null,
    };
  }

  public toJSON(): string {
    return JSON.stringify(instanceToPlain(this));
  }

  static fromJSON(json: string): Condition {
    return [plainToInstance(Condition, JSON.parse(json))].flat()[0];
  }
}

File: src-tauri/src/interface/mock/condition.rs
use crate::models::condition::Condition;
pub fn get_conditions() -> Vec<Condition> {
    vec![
        Condition {
            name: "15C".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(4.0),
        },
        Condition {
            name: "25C".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
        Condition {
            name: "Histamine".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
        Condition {
            name: "Tetracycline".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
    ]
}
pub fn get_filtered_conditions() -> Vec<Condition> {
    vec![
        Condition {
            name: "25C".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
        Condition {
            name: "Histamine".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
        Condition {
            name: "Tetracycline".to_string(),
            description: None,
            male_mating: Some(3),
            lethal: Some(false),
            female_sterile: Some(false),
            arrested: Some(false),
            maturation_days: Some(3.0),
        },
    ]
}
pub fn get_filtered_conditions_not_3_maturation_days() -> Vec<Condition> {
    vec![Condition {
        name: "15C".to_string(),
        description: None,
        male_mating: Some(3),
        lethal: Some(false),
        female_sterile: Some(false),
        arrested: Some(false),
        maturation_days: Some(4.0),
    }]
}
pub fn search_conditions_by_name() -> Vec<Condition> {
    vec![Condition {
        name: "Histamine".to_string(),
        description: None,
        male_mating: Some(3),
        lethal: Some(false),
        female_sterile: Some(false),
        arrested: Some(false),
        maturation_days: Some(3.0),
    }]
}
pub fn get_altering_conditions() -> Vec<Condition> {
    vec![Condition {
        name: "25C".to_string(),
        description: None,
        male_mating: Some(3),
        lethal: Some(false),
        female_sterile: Some(false),
        arrested: Some(false),
        maturation_days: Some(3.0),
    }]
}
